8353786: Migrate Vector API math library support to FFM API

iwanowww opened the pull request:
Migrate Vector API math library (SVML and SLEEF) linkage from native code (in JVM) to Java FFM API.

Since FFM API doesn't support vector calling conventions yet, migration affects only symbol lookup for now. But it still enables significant simplifications on JVM side.

The patch consists of the following parts:
  * on-demand symbol lookup in Java code replaces eager lookup from native code during JVM startup;
  * 2 new VM intrinsics for vector calls (support unary and binary shapes) (code separated from unary/binary vector operations);
  * new internal interface to query supported CPU ISA extensions (`jdk.incubator.vector.CPUFeatures`) used for CPU dispatching.

`java.lang.foreign` API is used to perform symbol lookup in vector math library, then the address is cached and fed into corresponding JVM intrinsic, so C2 can turn it into a direct vector call in generated code.

Once `java.lang.foreign` supports vectors & vector calling conventions, VM intrinsics can go away. 

Performance is on par with original implementation (tested with microbenchmarks on linux-x64 and macosx-aarch64).

Testing: hs-tier1 - hs-tier6, microbenchmarks (on linux-x64 and macosx-aarch64)

Thanks!

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8353786](https://bugs.openjdk.org/browse/JDK-8353786): Migrate Vector API math library support to FFM API (**Enhancement** - P3)


### Reviewers
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**) Review applies to [bb1a11db](https://git.openjdk.org/jdk/pull/24462/files/bb1a11dbc25dd65036acd313f7c25e91ee7305e4)
 * [Paul Sandoz](https://openjdk.org/census#psandoz) (@PaulSandoz - **Reviewer**) Review applies to [a288cbbf](https://git.openjdk.org/jdk/pull/24462/files/a288cbbfc2f17bf956c0e640f6e87d726e3cc264)
 * [Xiaohong Gong](https://openjdk.org/census#xgong) (@XiaohongGong - Committer) Review applies to [88eacc48](https://git.openjdk.org/jdk/pull/24462/files/88eacc48ef5b2eb40108a1cee6f892c9c0c5e6ce)
 * [Jorn Vernee](https://openjdk.org/census#jvernee) (@JornVernee - **Reviewer**) Review applies to [88eacc48](https://git.openjdk.org/jdk/pull/24462/files/88eacc48ef5b2eb40108a1cee6f892c9c0c5e6ce)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**) Review applies to [585312ae](https://git.openjdk.org/jdk/pull/24462/files/585312ae4ce114e13efef425634808fb16b3da29)
 * [Jatin Bhateja](https://openjdk.org/census#jbhateja) (@jatin-bhateja - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/24462/head:pull/24462` \
`$ git checkout pull/24462`

Update a local copy of the PR: \
`$ git checkout pull/24462` \
`$ git pull https://git.openjdk.org/jdk.git pull/24462/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 24462`

View PR using the GUI difftool: \
`$ git pr show -t 24462`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/24462.diff">https://git.openjdk.org/jdk/pull/24462.diff</a>

</details>
<details><summary>Using Webrev</summary>

[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/24462#issuecomment-2780154138)
</details>

liach commented:
Moving vector API library selection to Java code looks like a right step to me.

iwanowww commented:
/cc hotspot

minborg commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,311 @@
+    Entry<T> constructEntry(Operator op, int opc, VectorSpecies<E> vspecies, IntFunction<T> implSupplier) {
+        if (LIBRARY.isSupported(op, vspecies)) {
+            String symbol = LIBRARY.symbolName(op, vspecies);
+            MemorySegment addr = LOOKUP.find(symbol)
```
It is better to use `LOOKUP.findOrThrow()` because it does not require lambda creation.

iwanowww replied:
Thanks, changed as you suggested. I introduced a try-catch block instead.

vnkozlov commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/CPUFeatures.java:
```diff
@@ -0,0 +1,81 @@
+        return features.contains(feature.toLowerCase());
+    }
+
+    public static class X64 {
```
Should we create `src/jdk.incubator.vector/cpu/` for CPU specific information? As separate refactoring.

vnkozlov replied:
ok

vnkozlov commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,311 @@
+    }
+
+    /**
+     * Naming convention in SVML vector math library.
```
Does this library has code for all AVX configurations?

iwanowww replied:
Yes, there are 4 configurations (`-XX:UseAVX=[0..3]`) in total covered by SVML library.

vnkozlov replied:
Good.

vnkozlov reviewed (commented):
Few questions?

liach commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/CPUFeatures.java:
```diff
@@ -0,0 +1,81 @@
+    private static Set<String> getCPUFeatures() {
+        String featuresString = VectorSupport.getCPUFeatures();
+        debug(featuresString);
+        String[] features = featuresString.toLowerCase().split(", "); // ", " is used as a delimiter
```
Please use `toLowerCase(Locale.ROOT)`: if the system locale is turkish, `I` and dotless i are two letters, and the dotless i will fail in the subsequent `validateFeatures` assertion. Same for `hasFeature`.

iwanowww replied:
Good point. Fixed.

iwanowww commented:
In addition to addressing review feedback, there are 2 updates:
  * SVML: I overlooked that 64-bit vectors are covered by original implementation; fixed now;
  * JVM:  `features_string` to `cpu_info_string` renaming  uniformly across all platforms

XiaohongGong commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -101,31 +174,51 @@ static String precisionLevel(Operator op) {
+            }
+            if (vspecies.vectorBitSize() < 128) {
+                return false; // 64-bit vectors are not supported
+            }
```
Thanks for your refactor. It's really a good job!  

It seems float type support 64-bit vector operations before (see https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/stubGenerator_aarch64.cpp#L9835). Will this change the behavior of 64-bit float vector? Thanks!

iwanowww replied:
Thanks for taking a look. 

In the latest version, Float64Vector cases should be properly handled.

XiaohongGong replied:
Oh, yes. That's right! Thanks and sorry for my distraction.

PaulSandoz commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -510,6 +476,105 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
 }
 
 // public static
+// <V extends Vector<E>, E>
+// V unaryOp(long address, Class<? extends V> vClass, Class<E> elementType, int length,
+//           V v,
+//           UnaryOperation<V, ?> defaultImpl)
+//
+// public static
+// <V extends VectorPayload, E>
+// V binaryOp(long address, Class<? extends V> vClass, Class<E> elementType, int length,
+//            V v1, V v2,
+//            BinaryOperation<V, ?> defaultImpl)
```
`debugName` parameter is missing

XiaohongGong replied:
It seems the function should be updated to `libraryBinaryOp` and `libraryUnaryOp`?

iwanowww replied:
Fixed.

PaulSandoz commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -510,6 +476,105 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
+  address entry_addr = (address)entry->get_con();
+
+  const char* debug_name = "<unknown>";
+  const TypeInstPtr* debug_name_oop = gvn().type(argument(8))->isa_instptr();
```
Should that be:
```
  const TypeInstPtr* debug_name_oop = gvn().type(argument(6 + arity))->isa_instptr();
```
?
Placing the `debugName` parameter before the vector parameters makes it easier to reason about IMO.

XiaohongGong commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,334 @@
+                if (isRISCV64()) {
+                    return false; // FIXME: only MAX shapes are supported on RISC-V
+                }
+                if (isAARCH64() && vspecies.vectorBitSize() > 128) {
+                    return false; // FIXME: SVE support only for MAX shapes
+                }
```
SVE also supports operations for partial vector size, which means the vector size is smaller than MAX size. For example, if the max vector size of a SVE architecture is 512-bit, the FP vector operations with `VectorShape.S_256_BIT` are also supported. They are implemented with the same scalable math functions in SLEEF.

Hence, I think this check and the assertion in line-198 can be removed. Thanks!

XiaohongGong replied:
Please see the `addr` definition code in https://github.com/openjdk/jdk/blob/master/src/hotspot/share/opto/vectorIntrinsics.cpp#L1877 . If queried `addr` returns `nullptr` for 256-bit vectors, and the arch supports scalable vector, then the `addr` will be assigned to the scalable ones.

iwanowww replied:
Ah, ok. Thanks for the pointer. I aligned lookup logic with existing behavior. 

I'd like to double-check one things: is it fine to use scalable vector variants for fixed-sized vector shapes of smaller size without any explicit masking/stripping of upper vector part?

XiaohongGong replied:
Yes, it's fine for lanewise oeprations. For others like cross-lanes, stores and other operations, we will generate a predicate to strip the upper vector part. The op list that need the masking are listed here: https://github.com/openjdk/jdk/blob/master/src/hotspot/cpu/aarch64/aarch64_vector.ad#L280

PaulSandoz reviewed (approved):
Very nice work.

I am not too familiar with the format of CPU features on various platforms. Is the regex a best guess effort or something known and/or used by HotSpot too? If the former perhaps we might get tripped up by the assert in `CPUFeatures.validateFeatures` potentially rendering the Vector API unusable. Would it be better instead to drop and debug log?

iwanowww commented:
Both `CPUFeatures.validateFeatures` and the checks it performs are assertions, so it has to be explicitly enabled by `-esa`. The assert validates some assumptions the code has about the format of features string VM produces. And feature names are controlled by JVM (JVM assigns names itself). If there's a mismatch, something is really broken. Why do you think it makes sense to be defensive here and try to recover if the checks fail?

PaulSandoz commented:
> Both `CPUFeatures.validateFeatures` and the checks it performs are assertions, so it has to be explicitly enabled by `-esa`. The assert validates some assumptions the code has about the format of features string VM produces. And feature names are controlled by JVM (JVM assigns names itself). If there's a mismatch, something is really broken. Why do you think it makes sense to be defensive here and try to recover if the checks fail?

I was uncertain how the names were produced and if some untested hardware with additional CPU features could trigger an assert. Since as you say they are controlled by the JVM then indeed things are very broken if there is a mismatch.

XiaohongGong reviewed (approved):
Looks good to me. Thanks for your updating!

JornVernee reviewed (approved):
Very interesting! Looks mostly good to me. Left a few inline notes.

iwanowww commented on src/hotspot/share/runtime/abstract_vm_version.hpp:
```diff
@@ -58,6 +58,8 @@ class Abstract_VM_Version: AllStatic {
   static uint64_t _features;
   static const char* _features_string;
 
+  static const char* _cpu_info_string;
```
The intention is to align `_features_string` with `_features` which enumerates well-known CPU capabilities JVM manages. As of now, `_features_string` contains more information, so I introduced `_cpu_info_string` to keep it.

Speaking of `test/lib/jdk/test/whitebox/cpuinfo/CPUInfo.java`, the approach chosen there may be fine for a test library, but we need a more stable API between JVM and JDK.

Hamlin-Li replied:
I'm fine with this.
But it might be better to change the spliting regex of  `_features_string` in CPUFeatures.java to support riscv cpu features format.

iwanowww replied:
Ok, I pushed an update. Let me know what you think about it.

iwanowww commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/CPUFeatures.java:
```diff
@@ -0,0 +1,82 @@
+    public static Set<String> features() {
+        return features;
+    }
+}
```
Fixed.

iwanowww commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,323 @@
+        static String getDefaultName() {
+            return switch (StaticProperty.osArch()) {
+                case "amd64", "x86_64" -> SVML;
+                case "aarch64" -> SLEEF;
```
Goot catch, fixed.

iwanowww commented on src/hotspot/share/runtime/abstract_vm_version.cpp:
```diff
@@ -340,6 +341,19 @@ void Abstract_VM_Version::insert_features_names(char* buf, size_t buflen, const
+                                                         size_t features_offset) {
+  assert(features_offset <= cpu_info_string_len, "");
+  if (features_offset < cpu_info_string_len) {
+    assert(cpu_info_string[features_offset + 0] == ',', "");
```
Alternatively, it's fine for now to completely drop `extract_features_string` call on linux-riscv  (as on some other platforms) and fix it separately. Then `VectorSupport.getCPUFeatures()` returns empty string. `VectorMathLibrary` doesn't rely on `CPUFeatures` on RISC-V.

Let me know how you prefer to handle it.

Hamlin-Li replied:
I think we still need this or similar thing on riscv. Please check my new comments below about rvv extension on riscv.
On the other hand, it's also good to have it on riscv for consistency, and there is a log output of "cpu features" in VectorMathLibrary.java

iwanowww commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,323 @@
+            try {
+                MemorySegment addr = LOOKUP.findOrThrow(symbol);
+                debug("%s %s => 0x%016x\n", op, symbol, addr.address());
+                T impl = implSupplier.apply(opc); // TODO: should call the very same native implementation eventually (once FFM API supports vectors)
```
IMO an in-memory representation for vectors is preferred when it comes to FFM linker calling conventions. 512-bit vector requires 8 longs, so some of them will end up passed on stack for any non-trivial case. And with in-memory representation, VM can elide it once FFM linker stub is intrinsified.

iwanowww commented on src/hotspot/share/prims/vectorSupport.cpp:
```diff
@@ -655,23 +614,34 @@ JVM_ENTRY(jint, VectorSupport_GetMaxLaneCount(JNIEnv *env, jclass vsclazz, jobje
+  const char* features_string = VM_Version::features_string();
+  assert(features_string != nullptr, "missing cpu features info");
+
+  ThreadToNativeFromVM ttn(thread);
+  return env->NewStringUTF(features_string);
```
Fair enough.

iwanowww commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,323 @@
+
+    @ForceInline
+    /*package-private*/ static
+    <E, V extends Vector<E>>
```
I don't have a strong preference, but I kept it aligned with `unaryOp`/`binaryOp` intrinsics.

Hamlin-Li commented on src/jdk.incubator.vector/share/classes/jdk/incubator/vector/VectorMathLibrary.java:
```diff
@@ -0,0 +1,323 @@
+    V unaryMathOp(Unary op, int opc, VectorSpecies<E> vspecies,
+                  IntFunction<VectorSupport.UnaryOperation<V,?>> implSupplier,
+                  V v) {
+        var entry = lookup(op, opc, vspecies, implSupplier);
```
Seems there is another issue for riscv here.
If the rvv extension is not supported on the running machine, it will still generate the code using rvv, this should lead to a crash at runtime?

Hamlin-Li replied:
In previous code, we use `UseRVV` to detect if rvv extension is supported.

On the other hand, user can choose to disable UseRVV if they want even if rvv extension is supported on the running machine.

iwanowww replied:
Does the following check catch `UseRVV == false` case on RISC-V? 
```
public boolean isSupported(Operator op, VectorSpecies<?> vspecies) {
...
            int maxLaneCount = VectorSupport.getMaxLaneCount(vspecies.elementType());
            if (vspecies.length() > maxLaneCount) {
                return false; // lacking vector support
            }
...
```

Hamlin-Li replied:
Thank you for updating! Looks good for riscv. I have ran some basic tests for vector API, passed. I did not ran benchmark, as riscv & aarch64 share the same way to bridge from java to sleef.

> Does the following check catch UseRVV == false case on RISC-V?

Yes. If you don't mind, an explicit comment might be helpful. As to me "lacking vector support" here means the vector length is not large enough, but it's quite subjective, so you are on the call.

> FTR both VectorSupport.getMaxLaneCount() and CPUFeatures don't rely on raw list of ISA extensions CPU supports, but only those reported by the JVM. So, if some feature support is disabled on JVM side, it won't be reported by VM_Version and, hence, CPUFeatures.

I'm fine with this.

iwanowww replied:
Thanks, I added some clarifications in the comments.

Hamlin-Li reviewed (commented):
Found another possible issue on riscv.

jatin-bhateja commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -510,6 +476,107 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
+    char* buf = NEW_ARENA_ARRAY(C->comp_arena(), char, buflen);
+    debug_name = debug_name_oop->const_oop()->as_instance()->java_lang_String_str(buf, buflen);
+  }
+  Node* vcall = make_runtime_call(RC_VECTOR,
```
By generating an upfront CallLeafVectorNode, we may miss out on performing any GVN-style optimization for trigonometric identities like the following. do you think creating a macro node which can lazily be expanded to call node during macro expansion will help.
 
arcsin(sin(x)) => x
arccos(cos(x)) => x
sin(arcsin(x) => x
cos(arccos(x) => x

jatin-bhateja commented on src/hotspot/share/opto/vectorIntrinsics.cpp:
```diff
@@ -510,6 +476,107 @@ bool LibraryCallKit::inline_vector_nary_operation(int n) {
+    char* buf = NEW_ARENA_ARRAY(C->comp_arena(), char, buflen);
+    debug_name = debug_name_oop->const_oop()->as_instance()->java_lang_String_str(buf, buflen);
+  }
+  Node* vcall = make_runtime_call(RC_VECTOR,
```
By generating an upfront CallLeafVectorNode, we may miss out on performing any GVN-style optimization for trigonometric identities like the following. do you think creating a macro node which can lazily be expanded to call node during macro expansion will help.

arcsin(sin(x)) => x
arccos(cos(x)) => x
sin(arcsin(x) => x
cos(arccos(x) => x

jatin-bhateja replied:
I agree, this is a typical graph transform which cannot be applied currently because we are generating CallLeafVectorNode upfront during parsing, If we prevent intrinsification then compiler will attempt inlining, generating a much complex graph shape which may not be reducible.

iwanowww replied:
I don't see any insurmountable problems performing such transformations on chains of `CallLeafVector` nodes. But the missing piece is information about the algebraic properties of native functions JVM can't derive on its own.

RealFYang commented on src/hotspot/cpu/riscv/riscv.ad:
```diff
@@ -1944,7 +1944,7 @@ const RegMask* Matcher::predicate_reg_mask(void) {
 // Vector calling convention not yet implemented.
 bool Matcher::supports_vector_calling_convention(void) {
-  return EnableVectorSupport && UseVectorStubs;
+  return EnableVectorSupport;
```
You might want to remove the use of `UseVectorStubs` in `Matcher::vector_return_value` at L1951.
```
assert(EnableVectorSupport && UseVectorStubs, "sanity");
```

iwanowww replied:
Good catch. Fixed.

iwanowww commented:
Thanks for the reviews!

/integrate
